go/teams/fast.go

```diff
@@ -23,13 +34,81 @@ type FastTeamChainLoader struct {
 // from the server, and then the keys can be output in the result. Or, if you're doing a fast-load for
 // the purposes of loading parent teams to compute a name, then you can specify "seqnosNeeded" to say that certain
 // sigchain sequence numbers are neeeded unstubbed.
-func (f *FastTeamChainLoader) Load(m libkb.MetaContext, arg keybase1.FastLoadTeamArg) (res keybase1.FastLoadTeamRes, err error) {
-	return res, err
+func (f *FastTeamChainLoader) Load(m libkb.MetaContext, arg keybase1.FastTeamLoadArg) (res keybase1.FastTeamLoadRes, err error) {
+
+	defer m.CTrace(fmt.Sprintf("FastTeamChainLoader#Load(%+v)", arg), func() error { return err })()
+
+	flr, err := f.load(m, fastLoadArg{FastTeamLoadArg: arg})
+	if err != nil {
+		return res, err
+	}
+
+	res = flr.FastTeamLoadRes
+
+	res.Name, err = f.verifyTeamNameViaParentLoad(m, arg.Id, arg.Public, flr.unverifiedName, flr.upPointer)
+	if err != nil {
+		return res, err
+	}
+
+	return res, nil
 }
 
-// resolveIDToNameViaMerkelTree takes a team ID, and a pointer to a parent team's sigchain, and computes
+// verifyTeamNameViaParentLoad takes a team ID, and a pointer to a parent team's sigchain, and computes
 // the full resolved team name. If the pointer is null, we'll assume this is a root team and do the
 // verification via hash-comparison.
-func (f *FastTeamChainLoader) resolveIDToNameViaParentLoad(m libkb.MetaContext, id keybase1.TeamID, parent *keybase1.LinkTriple) (res keybase1.TeamName, err error) {
-	return res, err
+func (f *FastTeamChainLoader) verifyTeamNameViaParentLoad(m libkb.MetaContext, id keybase1.TeamID, isPublic bool, unverifiedName keybase1.TeamName, parent *keybase1.LinkTriple) (res keybase1.TeamName, err error) {
+
+	defer m.CTrace(fmt.Sprintf("FastTeamChainLoader#verifyTeamNameViaParentLoad(%s,%s)", id, unverifiedName), func() error { return err })()
+
+	if parent == nil {
+		if !unverifiedName.IsRootTeam() {
+			return res, NewBadNameError("expected a root team")
+		}
+		if !unverifiedName.ToTeamID(isPublic).Eq(id) {
+			return res, NewBadNameError("root team v. team ID mismatch")
+		}
+		return unverifiedName, nil
+	}
+
+	parentRes, err := f.load(m, fastLoadArg{
+		FastTeamLoadArg: keybase1.FastTeamLoadArg{
+			Id:     parent.Id,
+			Public: isPublic,
+		},
+		seqnosNeeded: []keybase1.Seqno{parent.Pair.Seqno},
+	})
+	if err != nil {
+		return res, err
+	}
+	downPointer, ok := parentRes.downPointers[parent.Pair.Seqno]
+	if !ok {
+		return res, NewBadNameError("down pointer not found in parent")
+	}
+	suffix := downPointer.NameComponent
+
+	parentName, err := f.verifyTeamNameViaParentLoad(m, parent.Id, isPublic, parentRes.unverifiedName, parentRes.upPointer)
```

Comment:
Would it be possible to have `load`'s contract be that it only returns verified names?

Reply:
That would shuffle everything around. I changed the return types upstream so that there's only one name in the return value. `fastLoadRes` has an unverified name. `FastTeamLoadRes` has a verified name.  The former no longer includes the latter.  Does that solve this issue?